### PR TITLE
Add cborplugin async send

### PR DESCRIPTION
**Pull request by mixmasala**
Refactor cborplugin so that plugins can send responses asynchronously from requests. This PR exposes SURBs to the CBORPlugin worker so that the plugin can respond at a later point in time, without blocking the mix server.

### CODE REVIEW

**Comment by mixmasala** [comment]
Question: how should the first-hop delay be chosen for SURB-ACK replies where the service plugin may use the SURB to reply any time within a valid Epoch?

**Review by mixmasala — COMMENTED** [review]
I've added comments where I think improvements should be made.

**Comment by mixmasala** [comment]
note, this PR merges https://github.com/katzenpost/katzenpost/pull/346

File: server/internal/provider/kaetzchen/cbor_plugins.go [hunk 1]
```diff
@@ -162,6 +164,13 @@ func (k *CBORPluginWorker) sendworker(pluginClient *cborplugin.Client) {
 						k.log.Debugf("%v: Failed to generate SURB-Reply: %v (%v)", pluginCap, r.ID, err)
 						continue
 					}
+					// Set the packet queue delay
```
Comment (david415) [inline server/internal/provider/kaetzchen/cbor_plugins.go:167]: doesn't seem fix the delay problem, you are applying the delay after the service processes the packet. we need the delay to happen before the service processes the payload.

**Review by david415 — CHANGES_REQUESTED** [review]
apply delay BEFORE service processing payload

**Comment by david415** [comment]
> Question: how should the first-hop delay be chosen for SURB-ACK replies where the service plugin may use the SURB to reply any time within a valid Epoch?

I think the first hop on the SURB replay (let's not call it an ACK) should be zero delay. The delay should be applied before the service processes the payload; because if the service processes the payload by making some kind of Internet traffic then the packet will be correlated with that network activity.

File: server/internal/packet/packet.go [hunk 2]
```diff
@@ -300,29 +297,35 @@ func NewPacketFromSURB(pkt *Packet, surb, payload []byte, geo *geo.Geometry) (*P
 	copy(nextHopCmd.ID[:], firstHop[:])
 	cmds = append(cmds, nextHopCmd)
 
-	nodeDelayCmd := new(commands.NodeDelay)
-	nodeDelayCmd.Delay = pkt.NodeDelay.Delay
-	cmds = append(cmds, nodeDelayCmd)
+	// pick a random delay to use for firstHop
```
[the remark this answers is not in the document]
Comment (david415) [reply server/internal/packet/packet.go]: We don't want to merge any PR that reduces the client's understanding of the RTT in my humble opinion; but we can probably fix this by applying a large fraction of the delay before the service processes the payload and optionally any remaining relay afterwards. We don't want service related network events to be linkable to the packets routed to the service Provider.

File: memspool/server/cmd/memspool/main.go [hunk 3]
```diff
@@ -107,20 +108,28 @@ func (s *spoolRequestHandler) OnCommand(cmd cborplugin.Command) (cborplugin.Comm
...
-		rawResp, err := resp.Marshal()
-		if err != nil {
-			s.log.Errorf("OnCommand called with invalid request")
-			return nil, err
+		if s.write == nil {
+			return errors.New("Plugin not registered")
 		}
-		return &cborplugin.Response{Payload: rawResp}, nil
+
+		go func() {
```
[the remark this answers is not in the document]
Comment (david415) [reply memspool/server/cmd/memspool/main.go:117]: are we even using memspool anymore?

File: server/cborplugin/client.go [hunk 4]
```diff
@@ -64,7 +64,9 @@ func (r *RequestFactory) Build() Command {
 
 // Response is the response received after sending a Request to the plugin.
 type Response struct {
+	ID      uint64
 	Payload []byte
+	SURB    []byte
 }
 
 // Marshal serializes Response
```
[the remark this answers is not in the document]
Comment (david415) [reply server/cborplugin/client.go:86]: yeah that sounds good if we have a use case for that. if not then i suggest saving that work until we actually need it.

File: memspool/server/cmd/memspool/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mixmasala) [reply memspool/server/cmd/memspool/main.go:117]: yes

**Comment by mixmasala** [comment]
Oops, re-reading this I see that we discussed adding the delay before processing the kaetzchen request.
https://github.com/katzenpost/katzenpost/issues/509

**Comment by mixmasala** [comment]
I think preserving the existing behavior is OK for now but we should address #509 somehow, in combination with #508

**Review by david415 — APPROVED** [review]

**Comment by david415** [comment]
@mixmasala we can merge this as soon as all the CI tests pass. the code quality is pretty good. i like how the reunion server was modified more than i like how the panda server was modified... because in the reunion case you at least keep track of the goroutine. but the code changes to the mix server itself are very good in my opinion.

**Review by david415 — APPROVED** [review]

**Review by david415 — APPROVED** [review]

**Pull request merged**